analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaEcommerceHelper.java

```diff
@@ -0,0 +1,115 @@
+      if (SEG_VIEWED_PRODUCT_CATEGORY.equalsIgnoreCase(track.event())) {
+        processProductCategory(track);
+      } else if (SEG_VIEWED_PRODUCT.equalsIgnoreCase(track.event())) {
+        processViewedProductName(track);
+        processProductCategory(track);
+      } else if (SEG_ADDED_PRODUCT.equalsIgnoreCase(track.event())) {
+        processAddedProductName(track);
+        processAddedProductCategory(track);
+      } else if (SEG_COMPLETED_ORDER.equalsIgnoreCase(track.event())) {
+        processCompletedOrderDiscount(track);
+      }
+    }
+  }
+
+  private void processCompletedOrderDiscount(TrackPayload track) {
+    String discountString = track.properties().getString(SEG_DISCOUNT_KEY);
+    if (!TextUtils.isEmpty(discountString)) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, discountString);
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    } else {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PURCHASE_DISCOUNT, "0");
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductCategory(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_CATEGORY, track.properties().category());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processAddedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_ADDED_TO_CART_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processViewedProductName(TrackPayload track) {
+    if (!TextUtils.isEmpty(track.properties().name())) {
+      Map<String, String> userAttributes = KahunaAnalytics.getUserAttributes();
+      userAttributes.put(LAST_PRODUCT_VIEWED_NAME, track.properties().name());
+      KahunaAnalytics.setUserAttributes(userAttributes);
+    }
+  }
+
+  private void processProductCategory(TrackPayload track) {
+
+    if (!TextUtils.isEmpty(track.properties().category())) {
+      Set<String> kahunaCategoriesSet
+              = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+        protected boolean removeEldestEntry(Entry<String, Boolean> eldest) {
+          return size() > MAX_CATEGORIES_VIEWED_ENTRIES;
+        }
+      });
+      if (KahunaAnalytics.getUserAttributes().containsKey(CATEGORIES_VIEWED)) {
+        String serializedCategories = KahunaAnalytics.getUserAttributes().get(CATEGORIES_VIEWED);
+        kahunaCategoriesSet.addAll(Arrays.asList(serializedCategories.split(",")));
```

Comment:
Great feedback. We debated on this internally and since we need to track a last category viewed of None if it isn't passed in by the developer, we need to track in the list of Categories Viewed as well to remain consistent. We think that while this is weird, it would reflect the correct state of the customer's information if they are tracking something without a category we'll reflect as None.